### ISSUE TITLE
Make "double-tap to activate" customizable on Android

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#14402]
Currently, whenever something is tap-able, Android gives the hint "double-tap to activate" and Flutter doesn't provide a way to override it. If you provide a custom hint via the Semantics widget, both hints (custom and default) are read out because we just add the custom hint to the regular text of the node (on iOS, the custom hint replaces the default hint as expected).

Android doesn't really provide a way to replace the default hint. Instead, starting with API 21 it can be customized via custom Accessibility Actions [1]. The description string given there will replace the "\<action\>" in "double-tap to \<action\>".

We should find a way to support that in Flutter.

[1] https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction.html#AccessibilityNodeInfo.AccessibilityAction(int, java.lang.CharSequence)

**Comment by alanrussian** [comment]
IIRC there's a support library equivalent that will make it easier to implement for all API levels.

**Comment by goderbauer** [comment]
@alanrussian Correct, the Android Support Libraries would provide this. Those should be used to implement this. We currently don't use them because that blows up our default binary size until https://github.com/flutter/flutter/issues/11099 is resolved.

**Comment by goderbauer** [comment]
The way we could implement this: 

* Add a `onTapHint`, `onLongPressHint` etc. String to the Semantics widget, which are then used on Android to generate custom hints via the API linked above, i.e. Android would concat that string with "double tap to..."
* The `hint` value would always be ignored on Android no matter if you provide `onTapHint` or not.
* On iOS, we would use the `hint` value, if provided, as `accessibilityHint`. If `hint` is not provided we generate a framework-side fallback hint out of the provided `onTapHint` Strings.

**Comment by Hixie** [comment]
Additionally:
* SemanticsProperties should assert if you provide a `tapHint` or `longPressHint` and don't provide a fallback `hint`.
* There should be a convenience API in the widgets library to autogenerate the `hint` for you based on strings provided in the `WidgetsLocalizations`.
* Semantics' default constructor should call that for you if you don't provide `hint` but o provide one of the other two.

I'm not sure I agree that we should always ignore `hint` on Android. If you provide a `hint` but nothing else, I would probably read it out like we do today.

**Comment by jonahwilliams** [comment]
Something related came up when working with  @DaveShuckerow on the reorder list semantics.  To make this accessible we would need to support actions on the local context menu to move the item up/down to end/start.  This behavior can be seen on the Android language selection.

I believe these actions need to be added via new AccessibilityAction, and without them a reorder list isn't accessible to switches or other input methods.

**Comment by jonahwilliams** [comment]
@goderbauer are doubleTapHint and longPressHint the only actions we would expect users to customize?  We might end up with each semantics node having a bunch of extra strings...

Another way to support this would be to extend the custom action API so that users can specify that a custom action actually overrides and existing action.  That would only work on Android however

**Comment by goderbauer** [comment]
We should probably support overriding the hint for all actions that android supports them on.

I like the idea of extending the custom actions to do this. As far as I know iOS does not support customizing this on the same granularity as Android does. On iOS, you can only customize the generic hint field, which we already expose.
